File: python/Trees/lowest_common_ancestor_binary_search_tree.py
# type: ignore
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        """
        Finds the lowest common ancestor (LCA) of two nodes in a BST.
        Approach:
        - Traverse from root, using BST property:
          - If both p and q are greater than current, go right.
          - If both are less, go left.
          - Otherwise, current is the LCA.
        Time Complexity: O(h) (h = height of tree)
        Space Complexity: O(1)
        """
        curr = root
        while curr:
            if p.val > curr.val and q.val > curr.val:
                curr = curr.right
            elif p.val < curr.val and q.val < curr.val:
                curr = curr.left
            else:
                return curr
        return None  # type: ignore

# -------------------
# Helper Functions for Testing
# -------------------
def build_bst_from_list(lst):
    if not lst:
        return None
    nodes = [TreeNode(val) if val is not None else None for val in lst]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()  # type: ignore
            if kids: node.right = kids.pop()  # type: ignore
    return root

def find_node(root, val):
    if not root:
        return None
    if root.val == val:
        return root
    left = find_node(root.left, val)
    if left:
        return left
    return find_node(root.right, val)

# -------------------
# Test Cases
# -------------------
def test_lowest_common_ancestor():
    s = Solution()
    # Test case 1: Example
    root = build_bst_from_list([6,2,8,1,4,7,9,None,None,3,5])
    p = find_node(root, 2)
    q = find_node(root, 8)
    assert s.lowestCommonAncestor(root, p, q).val == 6  # type: ignore
    p = find_node(root, 2)
    q = find_node(root, 4)
    assert s.lowestCommonAncestor(root, p, q).val == 2  # type: ignore
    # Test case 2: Single node
    root = build_bst_from_list([1])
    p = find_node(root, 1)
    q = find_node(root, 1)
    assert s.lowestCommonAncestor(root, p, q).val == 1  # type: ignore
    # Test case 3: LCA is root
    root = build_bst_from_list([5,3,6,2,4])
    p = find_node(root, 2)
    q = find_node(root, 6)
    assert s.lowestCommonAncestor(root, p, q).val == 5  # type: ignore
    print("All test cases passed!")

if __name__ == "__main__":
    test_lowest_common_ancestor()

File: C++/Trees/lowest_common_ancestor_binary_search_tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    /**
     * Finds the Lowest Common Ancestor (LCA) of two nodes in a Binary Search Tree.
     * 
     * Approach:
     * - Traverse starting from root.
     * - If both p and q are greater than current node, go to right child.
     * - If both are less than current node, go to left child.
     * - If they are on opposite sides or equal to current → current is LCA.
     *
     * Time Complexity: O(h), where h = height of the tree.
     * Space Complexity: O(1), iterative approach uses constant space.
     */
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        TreeNode* curr = root;

        while (curr) {
            if (p->val > curr->val && q->val > curr->val) {
                // Both nodes are in the right subtree
                curr = curr->right;
            }
            else if (p->val < curr->val && q->val < curr->val) {
                // Both nodes are in the left subtree
                curr = curr->left;
            }
            else {
                // Nodes are on different sides or one is equal → LCA
                return curr;
            }
        }

        return nullptr; // Should never reach here if both nodes are guaranteed to exist
    }
};

/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers.

using namespace std;
    ✅ Allows you to use standard library names (like cout, vector) without std:: prefix.

struct TreeNode {
    ✅ Defines the binary tree node structure.

int val;
    ✅ Value stored in the node.

TreeNode *left, *right;
    ✅ Pointers to the left and right child nodes.

TreeNode() : val(0), left(nullptr), right(nullptr) {}
    ✅ Default constructor: value=0, children=null.

TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    ✅ Constructor: initializes value with x, children=null.

TreeNode(int x, TreeNode *left, TreeNode *right)
    ✅ Constructor: initializes value with x, assigns left & right children.

class Solution {
    ✅ Defines the Solution class.

public:
    ✅ Public methods of the class.

TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q)
    ✅ Function to find the LCA of nodes p and q in the BST.

TreeNode* curr = root;
    ✅ Start traversal from the root.

while (curr) {
    ✅ Loop until current node becomes null.

if (p->val > curr->val && q->val > curr->val)
    ✅ Both nodes are greater than current → move right.

curr = curr->right;
    ✅ Move to right subtree.

else if (p->val < curr->val && q->val < curr->val)
    ✅ Both nodes are less than current → move left.

curr = curr->left;
    ✅ Move to left subtree.

else
    ✅ Nodes are on opposite sides or one matches → found LCA.

return curr;
    ✅ Return the lowest common ancestor.

return nullptr;
    ✅ Fallback: if not found (theoretically unreachable if inputs are valid).

*/
